microsoft/PowerToys
# Move KBM engine into separate process

## Summary of the Pull Request

**What is this about:**

**What is include in the PR:** 
- Move remapping logic from the module interface into a separate process
- Move corresponding tests into a separate project
- Dedicated logger folder for the new process
- Name of the new executable: `PowerToys.KeyboardManager.exe`
- Start KBM engine from the runner and set real-time priority for the process
- Exit the process if parent process(runner) exited
- New c++ file watcher. Implementation is based on WIN32 API

**TODO:**
- ~~Use the `Keyboard.ico` icon for `PowerToys.KeyboardManager.exe`~~
- ~~Add telemetry~~
- ~~Do not allow multiple instances of `PowerToys.KeyboardManager.exe`~~
- ~~Apply new log files structure to the project~~
- ~~When `Remap key window` or `Remap shortcut window` is opened remapings are suspended~~
- ~~Sync settings reading and remapping~~

**How does someone test / validate:** 
- Run PowerToys
- Try to edit key/shortcut remapping. Take into account, that `Press a key` windows does not work because low-level hook is deleted from the KBM module interface
- Verify that remapping are applied
- Check process priority. When running elevated priority should be real-time and when running as a user it should be high
- Try to restart PowerToys as administrator
- Disable/enable KBM

Future considerations:
- Rename projects

## Quality Checklist


## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.
